i/builtin: add gpio-chardev interface

ZeyadYasser opened the pull request:
This implements the `gpio-chardev` interface according to the SD129 spec which is currently hidden behind an experimental feature flag until kernel improvements to the `gpio-aggreagtor` lands.

Also, I had to extend the systemd backend to support specifying Wants/WantedBy/After/Before directives 44562bbde0bb6bba85a403de94a6f7e9c4b23f5c.

bboozzoo commented on interfaces/builtin/gpio_chardev.go:
```diff
@@ -0,0 +1,230 @@
+	return lines, nil
+}
+
+func validateLines(linesAttr string) error {
```
this could be made a helper in strutil, e.g.:

```
// RangeSpan represents a span of numbers inside a range. A span with
// equal Start and Stop describes a span of a single number.
type RangeSpan struct {
    Start, Stop uint
}

// Range of discrete numbers represented as set of non overlapping spans.
type Range struct {
    // Spans within the range, ordered by Start.
    Spans []RangeSpan 
}

// ParseRange parses a range represented as a string. The entries are 
// joining them with a comma: n[,m] or as a range: n-m or a combination
// of both, assuming the ranges do not overlap, e.g.: n,m,x-y.
func ParseRange(in str, expectedMax uint) (Range, error) {
    ...
}
```

and then you can reuse it in the helper. Separately I'd add something to the gadget package, or a subpackage there to convert range to a sequence of lines, which should be quite trivial at this point.

ZeyadYasser replied:
Thanks, this now could easily be shared with the helper command, about converting a range into a sequence of lines, I don't think we really need this as the kernel in fact accepts this exact format.

bboozzoo commented on interfaces/builtin/gpio_chardev.go:
```diff
@@ -0,0 +1,230 @@
+func (iface *gpioChardevInterface) BeforePrepareSlot(slot *snap.SlotInfo) error {
+	var sourceChip []string
+	// "source-chip" attribute is mandatory.
+	if err := slot.Attr("source-chip", &sourceChip); err != nil {
```
wort checking that the entries aren't empty strings, or duplicates

ZeyadYasser replied:
good point, thanks!

bboozzoo commented on interfaces/builtin/gpio_chardev.go:
```diff
@@ -0,0 +1,230 @@
+
+func (iface *gpioChardevInterface) UDevConnectedPlug(spec *udev.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
+	// Match exported virtual device on special udev tag set by snap-gpio-helper.
+	spec.TagDevice(fmt.Sprintf(`TAG=="snap_%s_interface_gpio_chardev_%s"`, slot.Snap().InstanceName(), slot.Name()))
```
Both instance and slot names could use characters which are not allowed in udev tags. And so they need to be changed a little, see udevTag() in interfaces/dev/spec.go. Maybe that helper should be moved to strutil, because you'll also need to access it in the helper.

ZeyadYasser replied:
I checked the validation for instance and slot names, and took a closer look at udevTag(). I am slightly confused, what characters that are allowed in instance and slot names but not in udev tags?

bboozzoo replied:
sorry I forgot that `-` are fine in tags and that we don't need to reverse the tag

pedronis reviewed (commented):
this looks reasonable to me

bboozzoo commented on strutil/range.go:
```diff
@@ -0,0 +1,118 @@
+}
+
+func parseRangeSpan(in string) (RangeSpan, error) {
+	hasNegativeStart := strings.HasPrefix(in, "-")
```
not at all, I discussed dropping it with @ZeyadYasser

ZeyadYasser replied:
Switched to using uints

ZeyadYasser commented on strutil/range.go:
```diff
@@ -0,0 +1,118 @@
+		}
+		r.Spans = append(r.Spans, s)
+	}
+	return r, nil
```
Good point, added sorting

bboozzoo commented on strutil/range.go:
```diff
@@ -0,0 +1,118 @@
+	Start, End uint
+}
+
+// Intersets checks if passed span intersects with this span.
```
```suggestion
// Intersects checks if passed span intersects with this span.
```

ZeyadYasser replied:
my bad, thanks!

bboozzoo commented on strutil/range.go:
```diff
@@ -0,0 +1,118 @@
+	Spans []RangeSpan
+}
+
+// Intersets checks if passed span intersects with this range of spans.
```
```suggestion
// Intersects checks if passed span intersects with this range of spans.
```

bboozzoo commented on strutil/range_test.go:
```diff
@@ -0,0 +1,65 @@
+
+	_, err = strutil.ParseRange("1,0-2")
+	c.Assert(err, ErrorMatches, `overlapping range span found "0-2"`)
+
```
You can add a couple more test cases here:
```suggestion
	_, err = strutil.ParseRange("1-3,0-2")
	c.Assert(err, ErrorMatches, `overlapping range span found "0-2"`)
	_, err = strutil.ParseRange("0-2,1-3")
	c.Assert(err, ErrorMatches, `overlapping range span found "1-3"`)
	_, err = strutil.ParseRange("0-2,2-3")
	c.Assert(err, ErrorMatches, `overlapping range span found "2-3"`)

```

bboozzoo commented on interfaces/builtin/gpio_chardev.go:
```diff
@@ -0,0 +1,202 @@
+
+// https://docs.kernel.org/userspace-api/gpio/chardev.html
+// https://docs.kernel.org/admin-guide/gpio/gpio-aggregator.html
+const gpioChardevSummary = `allows access to specific GPIO chardev lines`
```
We don't do that often/at all, but maybe you could add a TLDR summary of how the interface is expected to work:
```
The interface operates as follows:
- uses snap-gpio-helper to set up a virtual GPIO device exposing specific
  lines defined in the slot as character device node at 
  /dev/snap/gpio-chardev/<slot-snap>/<slot-name>
- sets up a symlink at /dev/snap/gpio-chardev/<plug-snap>/<plug-name>
  pointing at the virtual slot device
- slot devices are given a specific udev tag, which consumer rules match on
```

pedronis replied:
we don't do it because summary is meant to be a one liner. We can leave a comment though, and we will need good docs for this

bboozzoo replied:
sorry, I meant a TLDR summary in the comment, not the actual summary summary of the interface

bboozzoo commented on interfaces/builtin/gpio_chardev_test.go:
```diff
@@ -0,0 +1,277 @@
+`
+
+func (s *GpioChardevInterfaceSuite) SetUpTest(c *C) {
+	dirs.SetRootDir(c.MkDir())
```
the test should mock os release, otherwise the libexecdir paths won't match if executing on opensuse or fedora

ZeyadYasser commented on interfaces/builtin/gpio_chardev.go:
```diff
@@ -0,0 +1,202 @@
+		Type:            "oneshot",
+		RemainAfterExit: true,
+		// snap-gpio-helper export-chardev "<chip-labels>" "<lines>" "<gadget>" "<slot-name>"
+		ExecStart: fmt.Sprintf("/usr/lib/snapd/snap-gpio-helper export-chardev %q %q %q %q", strings.Join(sourceChip, ","), lines, snapName, slotName),
+		// snap-gpio-helper unexport-chardev "<chip-labels>" "<lines>" "<gadget>" "<slot-name>"
+		ExecStop: fmt.Sprintf("/usr/lib/snapd/snap-gpio-helper unexport-chardev %q %q %q %q", strings.Join(sourceChip, ","), lines, snapName, slotName),
```
great catch, thank you! this wouldn't have been caught until quite late

bboozzoo reviewed (approved):
LGTM

pedronis commented on interfaces/builtin/gpio_chardev.go:
```diff
@@ -0,0 +1,213 @@
+	return nil
+}
+
+// XXX: What should be the limit on max range.
```
@bboozzoo any suggestion here? is there a  limit in the kernel? otherwise we should just remove the XXX

bboozzoo replied:
we can set it to 512, which would still be very high and unlikely to ever be reached in practice.

pedronis reviewed (commented):
thanks, couple of small things, question for @bboozzoo

ZeyadYasser commented on strutil/range.go:
```diff
@@ -0,0 +1,118 @@
+	// Parse range e.g. 2-5
+	tokens := strings.SplitN(in, "-", 2)
+	if len(tokens) != 2 {
+		return RangeSpan{}, fmt.Errorf("invalid range span %q", in)
```
This check is not reachable since input that doesn't contain `-` will have already been parsed as a single number, and doing `SplitN` on a string that has `-` will always have exactly two tokens.

The check is redundant, but good to have in case `SplitN` was swapped to `Split` for any reason or by mistake.

ZeyadYasser commented:
I had to rebase to resolve conflicts

pedronis reviewed (approved):
thanks

ZeyadYasser commented on strutil/range.go:
```diff
@@ -0,0 +1,118 @@
+// Range of discrete numbers represented as set of non overlapping spans.
+type Range struct {
+	// Spans within the range, ordered by Start.
+	Spans []RangeSpan
```
Thanks, this is neater

zyga commented on interfaces/builtin/gpio_chardev.go:
```diff
@@ -0,0 +1,212 @@
+	plugName := slot.Name()
+	plugSnapName := plug.Snap().InstanceName()
+
+	target := fmt.Sprintf("/dev/snap/gpio-chardev/%s/%s", slotSnapName, slotName)
```
I'd really use `filepath.Join` here.

zyga commented on interfaces/builtin/gpio_chardev.go:
```diff
@@ -0,0 +1,212 @@
+		Type:            "oneshot",
+		RemainAfterExit: true,
+		ExecStart:       fmt.Sprintf("/bin/sh -c 'mkdir -p %q && ln -s %q %q'", filepath.Dir(symlink), target, symlink),
+		ExecStop:        fmt.Sprintf("/bin/sh -c 'rm -f %q'", symlink),
```
Do we need to run this via the shell?

ZeyadYasser replied:
not really, I was keeping it consistent with `ExecStart`, I can switch to `/bin/rm`

zyga commented on interfaces/builtin/gpio_chardev.go:
```diff
@@ -0,0 +1,212 @@
+	plugSnapName := plug.Snap().InstanceName()
+	snippet := "# Allow access to exported gpio chardev lines\n"
+	// Allow access to exported virtual slot device.
+	snippet += fmt.Sprintf("/dev/snap/gpio-chardev/%s/%s rwk,\n", slotSnapName, slot.Name())
```
We are computing this many times in the file. Can you use a helper that has the path exactly once.

ZeyadYasser replied:
Good point, I already did that in the helper command PR as a follow up

https://github.com/canonical/snapd/pull/15212/files#diff-10f4c2a34c9c6fe86ebdebad743920f2190b6af510ee25458ec839a7369d9e05

zyga commented on interfaces/builtin/gpio_chardev.go:
```diff
@@ -0,0 +1,212 @@
+	// Allow access to exported virtual slot device.
+	snippet += fmt.Sprintf("/dev/snap/gpio-chardev/%s/%s rwk,\n", slotSnapName, slot.Name())
+	// Allow access to plug-side symlink to exported virtual slot device.
+	snippet += fmt.Sprintf("/dev/snap/gpio-chardev/%s/ r,\n", plugSnapName)
```
I think we can fold this to `/{,*} r`

zyga commented on interfaces/systemd/service.go:
```diff
@@ -53,6 +57,18 @@ func (s *Service) String() string {
 		fmt.Fprintf(&buf, "ExecStop=%s\n", s.ExecStop)
 	}
+	if s.Wants != "" {
+		fmt.Fprintf(&buf, "Wants=%s\n", s.Wants)
```
Buf should have been a `strings.Builder` not `bytes.Buffer`

ZeyadYasser replied:
Good point, I think this should go into a followup PR, This one is already huge

ZeyadYasser commented on strutil/range.go:
```diff
@@ -0,0 +1,115 @@
+	"strings"
+)
+
+// RangeSpan represents a span of numbers inside a range. A span with
```
I don't have strong opinions on this, @bboozzoo @pedronis what do you think?

zyga commented on interfaces/builtin/gpio_chardev.go:
```diff
@@ -0,0 +1,212 @@
+	"github.com/snapcore/snapd/strutil"
+)
+
+// TODO: Snapd should validate the correctness of slot declarations
```
What's the plan to implement this after this PR lands?

ZeyadYasser replied:
Yes, It is tracked as https://warthogs.atlassian.net/browse/SNAPDENG-31777

zyga commented on interfaces/builtin/gpio_chardev.go:
```diff
@@ -0,0 +1,212 @@
+			dirs.DistroLibExecDir, strings.Join(sourceChip, ","), lines, snapName, slotName),
+		// snapd.gpio-chardev-setup.target is used for synchronization of
+		// app services and virtual device setup during boot.
+		WantedBy: "snapd.gpio-chardev-setup.target",
```
Will this behave correctly though? Will this affect things like server on a raspberry pi re-executing to new snapd?

ZeyadYasser replied:
You mean on classic, not core?

zyga reviewed (commented):
Mostly LGTM, my main remaining question is the TODO up top in `gpio_chardev.go`. I have some concerns about reexec on old snapd. Deferring my vote until those two are answered.

zyga reviewed (approved):
LGTM

ZeyadYasser commented:
For awareness, I found a stupid bug where the slot name was used as the plug name for the symlink generation, I fixed it in https://github.com/canonical/snapd/pull/15172/commits/b80228a350ec9a5d7d92d8aceed07c374b2b383f and updated the unit tests to catch it.
